Update to support elastic stack 5.x

shaidar opened the pull request:
### What are the relevant tickets?
[Issue#5](https://github.com/mitodl/elasticsearch-formula/issues/5)
### What's this PR do?
PR supports calling different versions of Elasticsearch along with installing Kibana, Nginx, and elastalert if use_elastic_stack is True. It also modified the sls file structure to line up with how we've been writing formulas going forward. 
### How should this be manually tests?
Clone repo and run vagrant up. Some automated tests have been written and included, however having issues with the testinfra module which needs be reviewed first before those tests end up working.

shaidar committed:
cleaning up and updating forumla

shaidar committed:
fixing problem with folder/file permissions and cleaning up

shaidar committed:
fixed error in starting service and cleaned up

shaidar committed:
changes after debugging and added some tests

blarghmatey commented on Vagrantfile:
```diff
@@ -10,11 +10,12 @@ Vagrant.configure(2) do |config|
     elasticsearch.vm.box = "debian/jessie64"
     elasticsearch.vm.provider "virtualbox" do |vb|
-      vb.memory = "2048"
+      vb.memory = "4096"
```
Is there a particular need to set the virtualbox size to 4G of RAM?

shaidar replied:
Kept on running into issues with the plugins and the jvm not having enough memory that bumping the value fixed it. I know it's not ideal, but guess if ES 5.x is allocating 2GB by default now, might be best to have the VM at 4.

blarghmatey commented on elasticsearch/configure.sls:
```diff
@@ -0,0 +1,109 @@
+
+{% for dir in (conf_dir, log_dir) %}
+{% if dir %}
+{{ dir }}:
```
I prefer to use descriptive IDs and always pass the `name` attribute explicitly. In this case it looks to be rather difficult to make a useful ID unless you do something like:
`create_directory_{{ dir|replace('/', '_') }}_to_ensure_proper_permissions`

blarghmatey commented on elasticsearch/configure.sls:
```diff
@@ -0,0 +1,109 @@
+{% set heap_size = heap_max if heap_max < 31744 else 31744 %}
+{% if elasticsearch.version == '5.x' %}
+config_jvm_options:
+  file.replace:
```
You can use a [`file.comment`](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.comment) here if all you are doing is commenting out the line.

shaidar replied:
file.comment appears to have been deprecated since 0.17.0 and file.replace is the recommendation.

blarghmatey commented on elasticsearch/elastalert/configure.sls:
```diff
@@ -0,0 +1,62 @@
+        - service: elastalert_service_running
+{% endfor %}
+
+create_elastalert_control_script:
```
I would argue that from this point down in this file belongs in `elastalert/install.sls` since it is part of the process of getting the service installed and running on the server. `configure.sls` is something that will be run somewhat frequently, whereas we should only ever need to define the init script and index once.

shaidar replied:
Makes sense. In my mind I was looking at .install as mainly installing the pkgs and everything else is part of .configure, but I do see your point on this and will make the change.

blarghmatey commented on elasticsearch/kibana/map.jinja:
```diff
@@ -1,7 +1,12 @@
+        'nginx_site_path': '/etc/nginx/sites-enabled',
         'config': {
-            'elasticsearch_url': 'http://localhost:9200',
+            'elasticsearch.url': 'http://localhost:9200',
```
What is your reasoning for replacing the `_` with a `.`?

shaidar replied:
Will need to add a condition here as it changed in 5.x:
https://www.elastic.co/guide/en/kibana/current/settings.html

blarghmatey commented on elasticsearch/plugins.sls:
```diff
@@ -16,6 +27,6 @@ plugin_configuration_for_{{ plugin.name }}:
         {{ plugin.config | yaml(False) | indent(8) }}
     - watch_in:
-        - service: elasticsearch_service
+        - service: elasticsearch
```
This is incorrect. The `elasticsearch_service` value is the id of the `service.running` state, not the name of the service (confusing, I know).

shaidar replied:
Good catch.

blarghmatey commented on elasticsearch/tests/test_elasticsearch_configure.sls:
```diff
@@ -0,0 +1,89 @@
+
+test_elasticsearch_config_file:
+	testinfra.file:
+		- name: {{ elasticsearch.conf_file }}
```
Indentation here is a bit off

blarghmatey commented on elasticsearch/tests/test_elasticsearch_install.sls:
```diff
@@ -0,0 +1,7 @@
+test_installed_{{ package }}:
+	testinfra.package:
+		- name: {{ package }}
+		- is_installed: True
```
Indentation here is a bit extreme. Each level should be indented 2 spaces.

blarghmatey commented on elasticsearch/tests/test_elasticsearch_service.sls:
```diff
@@ -0,0 +1,5 @@
+	testinfra.service:
+		- name: elasticsearch
+		- is_running: True
+		- is_enabled: True
```
Indentation. Should be 2 spaces per level.

blarghmatey commented on elasticsearch/repository.sls:
```diff
@@ -4,23 +4,49 @@
 
-{% for name, version in elasticsearch.products.items() %}
-configure_{{ name }}_package_repo:
+{% if elasticsearch.version == '5.x' %}
```
In some places you are comparing against the elasticsearch version, but in others you're comparing against `use_elastic_stack`. I would prefer to use the comparison against `use_elastic_stack` everywhere because it is a boolean and more future proof. In the not too distant future we might want to install 6.x and they will (hopefully) still be using the same package archive setup.

blarghmatey commented on elasticsearch/kibana/tests/test_kibana_service.sls:
```diff
@@ -0,0 +1,11 @@
+	testinfra.service:
+		- name: kibana
+		- is_running: True
+		- is_enabled: True
```
Indentation. 2 spaces per level.

blarghmatey commented on elasticsearch/kibana/install.sls:
```diff
@@ -0,0 +1,7 @@
+{% from "elasticsearch/kibana/map.jinja" import kibana with context %}
+
+install_kibana:
```
This should have a require statement for the package repo being configured. To do that you will need to:
```yaml
include:
  - elasticsearch.pkg_repo (or whatever the proper name is)
```
and then:
```yaml
install_kibana:
  pkg.installed:
    - require:
      - pkgrepo: <id of state to configure Kibana repo>
```

blarghmatey commented on elasticsearch/kibana/init.sls:
```diff
@@ -1,145 +1,8 @@
-    - reload_modules: True
-    - update: True
 
+{% if elasticsearch.use_elastic_stack %}
```
This shouldn't be conditional on whether the elastic stack is being used. Kibana is useful and usable against earlier versions of elasticsearch and we may not want to rely on the elastic stack repos to install Kibana.

I may be wrong in thinking this, so feel free to push back.

blarghmatey reviewed (changes requested):
This is a sizable PR so I'm sure I've missed a few things. I'm going to do another review of it, but this should give some things to work on in the meantime.

pdpinch commented:
@shaidar what's the status of this PR?

shaidar commented:
@pdpinch I believe there was an issue that's not captured in the comments when it comes to installing the right version of elasticsearch (old version for edx and then new version) and so I need to go back to this formula and refactor more code. I'll try and get that done tomorrow and get this PR merged.

shaidar committed:
Changes based on feedback

blarghmatey commented on elasticsearch/configure.sls:
```diff
@@ -0,0 +1,108 @@
+  file.replace:
+  - name: '{{ elasticsearch.conf_folder }}/jvm.options'
+  - pattern: '^-Xm[sx]\d\w'
+  - repl: '#-Xm[sx]\d\w'
```
This looks like it will replace the value with the contents of the regex rather than with any capture groups so that it will not only comment out the line, but eliminate the original string.

blarghmatey commented on elasticsearch/repository.sls:
```diff
@@ -1,26 +1,53 @@
         - pkgrepo: configure_elasticsearch_package_repo
 {% endif %}
 
-{% for name, version in elasticsearch.products.items() %}
```
We don't want to get rid of the for loop here because that is what allows us to support Kibana, Logstaash, etc. for pre 5.x releases.

blarghmatey commented on minion.conf:
```diff
@@ -0,0 +1,11 @@
+
+gitfs_remotes:
+  - https://github.com/mitodl/salt-extensions:
+      - root: extensions
```
What's the reason for pulling in the extensions repo here?

shaidar replied:
This file shouldn't even be part of this. Think it's generated when I ran the vagrantfile. Will remove it in next commit.

shaidar committed:
Changes based on feedback

blarghmatey commented on elasticsearch/kibana/install.sls:
```diff
@@ -1,7 +1,12 @@
\ No newline at end of file
+    - update: True
+    - require:
+      - pkgrepo: configure_elasticsearch_package_repo
```
Won't this be different depending on whether it's using elastic_stack or not?

shaidar replied:
Oh, the comment is a question and not a statement. I sense progress in my skills :-)
Based on my reading of it, I don't think so since the logic is placed in the configure_elasticsearch_package_repo for adding the appropriate repo based on whether elastic_stack is True or False

blarghmatey commented on elasticsearch/repository.sls:
```diff
@@ -1,18 +1,35 @@
+{% else %}
 {% for name, version in elasticsearch.products.items() %}
-configure_{{ name }}_package_repo:
+configure_elasticsearch_package_repo:
```
This line needs to stay the same since it is a for loop and will otherwise result in a duplicate state id error.

shaidar committed:
Changes based on feedback

shaidar committed:
Changes based on feedback

blarghmatey committed:
Fixing Testinfra tests

blarghmatey committed:
Fixed installation of Elastalert

- Added state to remove `python-cryptography` OS package in favor of
  dependency from Elastalert pip install
- Fixed invalid indentation in Kibana tests
